bundles/org.openhab.binding.bluetooth/src/main/java/org/openhab/binding/bluetooth/discovery/internal/BluetoothDiscoveryProcess.java

```diff
@@ -0,0 +1,315 @@
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.bluetooth.BluetoothAddress;
+import org.openhab.binding.bluetooth.BluetoothBindingConstants;
+import org.openhab.binding.bluetooth.BluetoothCharacteristic;
+import org.openhab.binding.bluetooth.BluetoothCharacteristic.GattCharacteristic;
+import org.openhab.binding.bluetooth.BluetoothCompanyIdentifiers;
+import org.openhab.binding.bluetooth.BluetoothCompletionStatus;
+import org.openhab.binding.bluetooth.BluetoothDescriptor;
+import org.openhab.binding.bluetooth.BluetoothDevice;
+import org.openhab.binding.bluetooth.BluetoothDevice.ConnectionState;
+import org.openhab.binding.bluetooth.BluetoothDeviceListener;
+import org.openhab.binding.bluetooth.discovery.BluetoothDiscoveryParticipant;
+import org.openhab.binding.bluetooth.notification.BluetoothConnectionStatusNotification;
+import org.openhab.binding.bluetooth.notification.BluetoothScanNotification;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link BluetoothDiscoveryProcess} does the work of creating a DiscoveryResult from a set of
+ * {@link BluetoothDisocveryParticipant}s
+ *
+ * @author Connor Petty - Initial Contribution
+ */
+@NonNullByDefault
+public class BluetoothDiscoveryProcess implements Supplier<DiscoveryResult>, BluetoothDeviceListener {
+
+    private static final int DISCOVERY_TTL = 300;
+
+    private final Logger logger = LoggerFactory.getLogger(BluetoothDiscoveryProcess.class);
+
+    private final Lock serviceDiscoveryLock = new ReentrantLock();
+    private final Condition connectionCondition = serviceDiscoveryLock.newCondition();
+    private final Condition serviceDiscoveryCondition = serviceDiscoveryLock.newCondition();
+    private final Condition nameDiscoveryCondition = serviceDiscoveryLock.newCondition();
+
+    private final Collection<BluetoothDiscoveryParticipant> participants;
+    private final BluetoothDevice device;
+
+    private volatile boolean servicesDiscovered = false;
+
+    public BluetoothDiscoveryProcess(BluetoothDevice device, Collection<BluetoothDiscoveryParticipant> participants) {
+        this.participants = participants;
+        this.device = device;
+    }
+
+    @Override
+    public DiscoveryResult get() {
+        // first see if any of the participants that don't require a connection recognize this device
+        List<BluetoothDiscoveryParticipant> connectionParticipants = new ArrayList<>();
+        for (BluetoothDiscoveryParticipant participant : participants) {
+            if (participant.requiresConnection(device)) {
+                connectionParticipants.add(participant);
+                continue;
+            }
+            try {
+                DiscoveryResult result = participant.createResult(device);
+                if (result != null) {
+                    return result;
+                }
+            } catch (RuntimeException e) {
```

Comment:
What can throw an exception here? I did look at `BlukiiDiscoveryParticipant? and `RuviiTagDiscoveryParticipant` and both require `device` to be non-null, but NPE seem to be the only exception possible. This should IMO be solved by proper null-checks (and annotations, which seem to be wrong) and not by catching `RuntimeException`.

Reply:
This is more or less legacy code that was copied from DiscoveryService.
But my understanding of its purpose is to prevent an error in any one DiscoveryParticipant from prevent discovery from another DiscoveryParticipant or the creation of a default discovery. You are correct that there is no reasonable exception that would be thrown by the current codebase. I think catching runtime exceptions here is useful for future-proofing the eventual creation of potentially complex DiscoveryParticipants which may interact with more than just the bluetooth device and thus may end up causing unintended errors during discovery.

Whether or not you agree with that sentiment is entirely up to you.